### ISSUE TITLE
BCELoss results in autocast CUDA warning

### ISSUE THREAD

**Issue by as51340** [pytorch/pytorch#83733]
### 🐛 Describe the bug

Hi folks, BCELoss on calling *backward* method creates warning UserWarning: User provided device_type of 'cuda', but CUDA is not available. Disabling   warnings.warn('User provided device_type of \'cuda\', but CUDA is not available. Disabling'). The code works normally but it is very irritant. torch/cuda/amp/autocast_mode.py creates a warning because it gets called with `device_type` "cuda". I can get around by commenting line 199 but that is not the solution. 

### Versions

o CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.22.2
[pip3] numpydoc==1.4.0
[pip3] torch==1.12.1+cpu
[pip3] torchaudio==0.12.1+cpu
[pip3] torchvision==0.13.1+cpu
[conda] Could not collect

**Comment by ptrblck** [comment]
The error is raised by a wrong usage of `torch.cuda.amp.autocast` or `torch.amp.autocast` with a wrong `device_type`.
Currently you can use two different approaches either via:
```python
with torch.cuda.amp.autocast():
    # your code
``` 
or the newer (and more general API) via:
```python
with torch.amp.autocast(device_type=device):
    # your code
```
The former approach will call the latter internally with `device_type="cuda"` as seen [here](https://github.com/pytorch/pytorch/blob/1fa9a377d01ba8e1a0b65cc2d05ed8a2d53a89f2/torch/cuda/amp/autocast_mode.py#L19-L26).

If you do not have a GPU or CUDA runtime installed, you want to use the CPU mixed-precision implementation and should thus use the latter approach with `device_type="cpu"`.

I don't see any `autocast` usage in your code so assume that DGL is explicitly using `torch.cuda.amp.autocast` without checking if a GPU or CUDA runtime is available.
Here is a small code snippet to reproduce the issue which you can run using the CPU-only wheels:
```python
import torch
import torch.nn as nn

device = 'cpu'
model = nn.Linear(10, 10).to(device)
x = torch.randn(1, 10).to(device)

#with torch.cuda.amp.autocast(): # warning
#with torch.amp.autocast(device_type='cuda'): # same warning
with torch.amp.autocast(device_type=device): # works on CPU
    out = model(x)

print(out.dtype)
```

I thus think that DGL should fix the `autocast` usage and don't see any issue in PyTorch since the user (DGL in this case) is explicitly asking for the CUDA backend in `autocast` in a CPU-only build.

**Comment by ptrblck** [comment]
Also, take this with a grain of salt:
> I thus think that DGL should fix the autocast usage

as I don't know if DGL even wants to support CPU mixed-precision training. 
So the warning and disablement of `autocast` might be by design.

Based on [these docs](https://github.com/dmlc/dgl/blob/6e1be69a84ba3e17e8e4db3c3768448f3620ecf4/docs/source/guide/mixed_precision.rst#end-to-end-mixed-precision-training) only the CUDA backend is mentioned, but this might also be due to historical reasons, since it was available before the CPU backend was added.

Adding @nv-dlasalle in case he is more familiar with DGL's amp support.

**Comment by nv-dlasalle** [comment]
DGL shouldn't be calling autocast at all in the above code.

@as51340 Can you include what version of DGL you're using? Can you also include a more complete example to reproduce this error?

**Comment by as51340** [comment]
DGL version is `0.9.0`. Here is the code `https://gist.github.com/as51340/78300bb57d079e8ec85a493af207fc8e`,

**Comment by ptrblck** [comment]
The code is unfortunately not executable and I also don't see any usage of `torch.amp`, so can't guess what might be causing the error.

**Comment by as51340** [comment]
Hi, this is runnable example that should reproduce the error: `https://gist.github.com/as51340/69691482e2a204e2a8e12c07954e0553`

**Comment by tg-bill** [comment]
I got the same warning in DGL 0.9.0. The warning was thrown when running `loss.backward()`. Do you have any suggestion on quick fixes rather than suppressing the warning?

**Comment by nv-dlasalle** [comment]
After diving deeper into the stack traces here, this looks to be a DGL issue. I opened https://github.com/dmlc/dgl/issues/4593 there to track this.
